src/include/index/ivf_pq_index.h

```diff
@@ -1244,16 +1281,141 @@ class ivf_pq_index {
         std::move(*generate_query_to_pq_centroid_distance_tables<
                   Q,
                   ColMajorMatrix<float>>(query_vectors));
-    return detail::ivf::query_finite_ram(
-        *partitioned_pq_vectors,
-        query_to_pq_centroid_distance_tables,
-        active_queries,
-        k_nn,
-        upper_bound,
-        num_threads_,
-        make_pq_distance_query_to_pq_centroid_distance_tables<
-            std::span<float>,
-            decltype(pq_storage_type{}[0])>());
+    size_t k_initial = static_cast<size_t>(k_nn * k_factor);
+    auto&& [initial_distances, initial_ids, initial_indices] =
+        detail::ivf::query_finite_ram(
+            *partitioned_pq_vectors,
+            query_to_pq_centroid_distance_tables,
+            active_queries,
+            k_initial,
+            upper_bound,
+            num_threads_,
+            make_pq_distance_query_to_pq_centroid_distance_tables<
+                std::span<float>,
+                decltype(pq_storage_type{}[0])>());
+    return re_rank(
+        std::move(initial_distances),
+        std::move(initial_ids),
+        std::move(initial_indices),
+        query_vectors,
+        k_initial,
+        k_nn);
+  }
+
+  auto re_rank(
```

Comment:
nit: I would prefer `rerank` or `rerank_results`

Reply:
Sure, done. Note `rerank` is not actually a word in the dictionary, but it does make it easier to write, and people know what it means anyways.